plugins/gstperf.c

```diff
@@ -401,16 +422,91 @@ gst_perf_cpu_get_load (GstPerf * perf, guint32 * cpu_load)
   } else {
     *cpu_load = 0;
   }
+
   /*Remember the total and idle CPU for the next check */
   perf->prev_cpu_total = total;
   perf->prev_cpu_idle = idle;
+
+  return TRUE;
+
+cpu_failed:
+  GST_ERROR_OBJECT (perf, "Failed to get the CPU load");
+  return FALSE;
+}
+
+#elif IS_MACOSX
+static gboolean
+gst_perf_cpu_get_load_macosx (GstPerf * perf, guint32 * cpu_load)
+{
+  guint32 idle = 0;
+  guint32 total = 0;
+  guint32 diff_total, diff_idle;
+  host_cpu_load_info_data_t cpuinfo;
+  mach_msg_type_number_t count = HOST_CPU_LOAD_INFO_COUNT;
+
+  g_return_val_if_fail (perf, FALSE);
+  g_return_val_if_fail (cpu_load, FALSE);
+
+  /* Default value in case of failure */
+  *cpu_load = -1;
+
+  if (host_statistics (mach_host_self (), HOST_CPU_LOAD_INFO,
+          (host_info_t) & cpuinfo, &count) == KERN_SUCCESS) {
+    for (int i = 0; i < CPU_STATE_MAX; i++) {
+      total += cpuinfo.cpu_ticks[i];
+    }
+    idle = cpuinfo.cpu_ticks[CPU_STATE_IDLE];
+  } else {
+    goto cpu_failed;
+  }
+
+  /*Calculate the CPU usage since last time we checked */
+  diff_idle = idle - perf->prev_cpu_idle;
+  diff_total = total - perf->prev_cpu_total;
+  if (diff_total) {
+    /*Get a rounded result */
+    *cpu_load = (1000 * (diff_total - diff_idle) / diff_total + 5) / 10;
```

Comment:
Will `diff_total` be always > than diff_idle? if not, cpu_load will be negative and `cpu_load` is unsigned int

Reply:
In theory yes, `total` time includes `idle` time in it.